Add page on formal verification for smart contracts (#6394)
ethereum/ethereum-org-website: src/content/developers/docs/smart-contracts/formal-verification/index.md

## What is formal verification? {#what-is-formal-verification}

Formal verification is used to evaluate the _correctness_ of software programs and confirm compliance with predefined specifications using formal methods of mathematics. [More on formal verification](https://en.wikipedia.org/wiki/Formal_verification).

In the context of Ethereum, formal verification can prove that a smart contract will only implement the desired business logic.

### Why formal verification is necessary {#why-formal-verification-is-necessary}

While manual code audits are great, human auditors cannot possibly cover every possible vulnerability. Also, an audit may reveal bugs in contract code, but it cannot confirm the absence of vulnerabilities.

Formal verification is grounded in [formal methods](https://en.wikipedia.org/wiki/Formal_methods), which use mathematically rigorous methods for specifying, developing, and verifying software. Formal verification allows developers to mathematically model software programs and create formalized specifications, making testing more precise and robust.

> Specifications are properties or used to describe the intended behavior of a smart contract. Formal verification can prove that a contract behaves exactly as described in the specification. Unlike testing based on informal assumptions, formal verification provides mathematical proof of a smart contract's correctness.

Reviewer: > Specifications are properties or used to

This sentence doesn't make sense to me. Should it be "Specifications are properties used to", or something else?

Author: It's a typo. The sentence should read "specifications are properties used to...". I'll edit the content to fix it.

## How does formal verification work? {#how-does-formal-verification-work}

Formal verification of Ethereum smart contracts can either be executed at the language level or bytecode level or both. Here’s a quick overview of both approaches:

### High-level verification {#high-level-verification}

> Source code for Ethereum smart contracts is usually [written in high-level programming languages](/developers/docs/smart-contracts/languages/), such as Solidity, Vyper, or Yul. The source code defines the business logic, which determines how users can interact with the contract and how the contrant should behave in response to certain inputs.

Reviewer: ```suggestion
Source code for Ethereum smart contracts is usually [written in high-level programming languages](/developers/docs/smart-contracts/languages/), such as Solidity, Vyper, or Yul. The source code defines the business logic, which determines how users can interact with the contract and how the contract should behave in response to certain inputs.
```

In the context of Ethereum, formal verification can prove that a smart contract will only implement the desired business logic.

### Why formal verification is necessary {#why-formal-verification-is-necessary}

While manual code audits are great, human auditors cannot possibly cover every possible vulnerability. Also, an audit may reveal bugs in contract code, but it cannot confirm the absence of vulnerabilities.

Formal verification is grounded in [formal methods](https://en.wikipedia.org/wiki/Formal_methods), which use mathematically rigorous methods for specifying, developing, and verifying software. Formal verification allows developers to mathematically model software programs and create formalized specifications, making testing more precise and robust.

Specifications are a collection of properties used to describe the intended behavior of a smart contract. Formal verification can prove that a contract behaves exactly as described in the specification. Unlike testing based on informal assumptions, formal verification provides mathematical proof of a smart contract's correctness.

> Formal verification is critical for smart contracts because minor vulnerabilities can lead to massive losses. For example, the [Parity Multisig wallet hack](https://www.freecodecamp.org/news/a-hacker-stole-31m-of-ether-how-it-happened-and-what-it-means-for-ethereum-9e5dc29e33ce/amp/) was the result of an unintended function that allowed the hacker to override owner settings and drain the contract of up to $31 million in ETH. 

Reviewer 2: Curious, would this have helped if the contract was formally verified?

Author: The consensus is that formal verification can help detect unintended code executions in smart contracts, so formal verification would probably have helped in the Parity case. More to the point, the Parity hack was mentioned in a couple of academic research and articles on formal verification, so I thought it'd be a good addition.

The actual verification at this stage consists of checking the source code and confirming that it satisfies all specifications and requirements. Therefore, high-level verification gives developers a high degree of confidence that the source code has no bugs.

### Low-level verification {#low-level-verification}

Smart contracts written in high-level languages are  compiled into **bytecode**—low-level machine instructions that the Ethereum Virtual Machine (EVM) understands. Source code may undergo subtle changes during compilation, which may cause the contract to execute operations not defined in high-level specifications.

Formal verification can help determine the correctness of a contract's compiled bytecode. This requires creating bytecode-based, low-level specifications that define how a smart contract should behave in the EVM environment.

Low-level formal verification uses techniques like static analysis and fuzzing to ensure that contract bytecode satisfies all properties. This assures developers that the business logic specified in the source code is what will run in the context of the EVM.

> ## Advantages and disadvantages of formal verification {#advantages-and-disadvantages-of-formal-verification}

Reviewer 2: Would break this header up... felt a little awkward reading and having the numbered list go 1. 2. 3. 1. 2. 3

```suggestion
## Advantages of formal verification {#advantages-of-formal-verification}
```

Formal verification is useful for the following reasons:

### 1. Comprehensiveness {#comprehensiveness}

Formal verification uses automated tests, which can be run many times over. A formal verification assessment will likely cover the entire contract, reducing edge cases and hidden vulnerabilities.

### 2. Rigorous testing {#rigorous-testing}

Although effective, manual audits have some elements of human error and uncertainty—which can have costly consequences in cases. Formal verification approaches contract auditing in a formal and systematic manner that reduces ambiguity to the barest minimum.

Formal verification transforms informal specifications into properties grounded in formal logic. This makes it easier to test assertions and evaluate the correctness of smart contract programs.

### 3. Faster development cycle {#faster-development-cycle}

Automated formal verification can be more efficient than line-by-line analysis of smart contract code. Formal verification approaches cover a larger surface area in a short timeframe. This can speed up the testing and development cycle for developers.

> Although formal verification is great for testing smart contracts, it has some disadvantages:

Reviewer 2: ```suggestion
## Disadvantages of formal verification {#disadvantages-of-formal-verification}

Although formal verification is great for testing smart contracts, it has some disadvantages:
```

### 1. Difficulty in creating formal specifications {#difficulty-in-creating-formal-specifications}

A prerequisite for formal verification is creating formal specifications, which can take a lot of time and effort. This will likely be the most difficult part of using formal verification for smart contracts.

> ### 2. False positives {#false-positives}

Reviewer 2: The description below would be a false negative (if the test _falsely_ gave a _negative_ test result for a buggy contract). A false positive would be if it flagged a bug when there wasn't one. 

```suggestion
### 2. False negatives {#false-negatives}
```

---
title: Formal verification of smart contracts
description: An overview of formal verification techniques for Ethereum smart contracts
lang: en
sidebar: true
---

[Smart contracts](/developers/docs/smart-contracts/) are making it possible to create decentralized, trustless, and robust applications that introduce new use-cases and unlock value for users. Because smart contracts handle large amounts of value, security is a critical consideration for developers.

Reliability is a highly desired quality of smart contracts, and both users and developers want assurances that a smart contract will work as intended. This is important because code in the [Ethereum Virtual Machine (EVM)](/developers/docs/evm/) is immutable, which rules out the possibility of fixing bugs or retrieving funds lost to malicious attacks.

> Formal verification is one of the recommended techniques for [testing smart contracts](/developers/docs/smart-contracts/testing/) and improving [smart contract security](/developers/docs/smart-contracts/security/). Formal verification involves testing a program (a smart contract in this case) to ensure it behaves as expected during execution.

Reviewer 3: > Formal verification involves testing a program

This part isn't really precise, since FV is not about testing. I would actually recommend removing this line completely, since line 16 below already introduces FV well.

Author: Thanks for pointing it out. I mixed up testing contracts (to check results) and evaluating contracts to see if they satisfy specific criteria (which is what I believe FV does). I'm working on rewriting certain page of the page to remove the inaccuracy.

## What is formal verification? {#what-is-formal-verification}

Formal verification is used to evaluate the _correctness_ of software programs and confirm compliance with predefined specifications using formal methods of mathematics. [More on formal verification](https://en.wikipedia.org/wiki/Formal_verification).

> In the context of Ethereum, formal verification can prove that a smart contract will only implement the desired business logic.

Reviewer 3: ```suggestion
In the context of Ethereum, formal verification usually attempts to prove that a smart contract has certain properties, often called invariants. If these properties are violated, attackers can cause harm to protocols and users.
```

Formal verification can help verify the correctness of a contract's source code. This requires transforming the business logic into a set of formalized, high-level specifications. These specifications must adequately and comprehensively express the desired functionality of the smart contract.

The actual verification at this stage consists of checking the source code and confirming that it satisfies all specifications and requirements. Therefore, high-level verification gives developers a high degree of confidence that the source code has no bugs.

### Low-level verification {#low-level-verification}

Smart contracts written in high-level languages are compiled into **bytecode**—low-level machine instructions that the Ethereum Virtual Machine (EVM) understands. Source code may undergo subtle changes during compilation, which may cause the contract to execute operations not defined in high-level specifications.

Formal verification can help determine the correctness of a contract's compiled bytecode. This requires creating bytecode-based, low-level specifications that define how a smart contract should behave in the EVM environment.

> Low-level formal verification uses techniques like static analysis and fuzzing to ensure that contract bytecode satisfies all properties. This assures developers that the business logic specified in the source code is what will run in the context of the EVM.

Reviewer 3: This is not 100% precise. There are also several tools that apply symbolic execution and model checking to EVM verification, which are FV techniques.

Author: Yeah. I did a bit more research and found some more FV techniques (e.g., symbolic execution, model checking, and theorem proving), which are not mentioned in the page. I'll add those details while reworking the page, though.

## Advantages of formal verification {#advantages-of-formal-verification}

Formal verification is useful for the following reasons:

### 1. Comprehensiveness {#comprehensiveness}

> Formal verification uses automated tests, which can be run many times over. A formal verification assessment will likely cover the entire contract, reducing edge cases and hidden vulnerabilities.

Reviewer 3: Not precise, FV doesn't use tests. It mathematically abstracts the software behavior in a way that specialized procedures can guarantee correctness for every possible input case, *without* having to exhaustively go through each one of them.

Author: Just figured this out, too, after reading your comment and researching existing literature. I'll add this to the page while editing. Your comment may also be useful in highlighting the difference between FV and regular testing (i.e., that regular testing can only cover a small number of input cases).

- [GitHub](https://github.com/crytic/echidna/)

**Harvey - _Automated fuzzing tool useful for detecting property violations in smart contract code._**

- [Website](https://consensys.net/diligence/fuzzing/)

### Code analysis {#code-analysis}

**Oyente - _Static analysis tool for analyzing vulnerabilities in EVM bytecode with symbolic execution._**

- [GitHub](https://github.com/melonproject/oyente)

**Mythril - _EVM bytecode assessment tool for detecting contract vulnerabilities using taint analysis, concolic analysis, and control flow checking._**

- [GitHub](https://github.com/ConsenSys/mythril-classic)
- [Documentation](https://mythril-classic.readthedocs.io/en/master/about.html)

**Manticore _- A tool for analyzing EVM bytecode analysis tool based on symbolic execution._**

- [GitHub](https://github.com/trailofbits/manticore)
- [Documentation](https://github.com/trailofbits/manticore/wiki)

**Securify** **_- A security analyzer for analyzing Solidity source code in Ethereum smart contracts and detecting property violations._**
- [GitHub](https://github.com/eth-sri/securify2)

> ### Formal verification projects in Ethereum {#formal-verification-projects-in-ethereum}

Reviewer 3: What is the exact meaning of this section? The tools in the previous section also fit under the "eth fv" umbrella.

Author: I was aiming to highlight tools that provide a more comprehensive framework for formal verification. For example, the KEVM framework was the most visible project I found that was developed specifically for formal verification. But if you want, I could have them listed together under a "Formal verification tools for Ethereum smart contracts" heading.

**hevm - _An implementation of the Ethereum virtual machine (EVM) made specifically for symbolic execution, unit testing and debugging of smart contracts._**
- [GitHub](https://github.com/dapphub/dapptools/tree/master/src/hevm)

**KEVM - _A correct-by-construction implementation of the Ethereum Virtual Machine (EVM) specified in the K Framework for implementing formal verification._**
- [GitHub](https://github.com/runtimeverification/evm-semantics)

> - [Documentation](https://jellopaper.org/)

Reviewer 3: There are quite a few common tools missing in this list. If the goal is to list only the "most used" or "most stable" ones, I'd have a few suggestions of tools to remove and to add.

Reviewer 4: @leonardoalt what tools would you recommend adding? I'd default to you on which to keep and/or remove.

Also happy to default to you on ordering... if it makes sense to put the most used tools at the top for visibility then let's do that.

Author: The idea was mostly to highlight most-used and maintained tools. For instance, MAIAN hasn't been maintained for a while and the KLab repo has been archived (I took most of the tools from your listing). That said, any suggestions on what tools to add or remove are welcome.

Thus, we can think of a formal specification as a collection of statements written in a formal language that describe the intended execution of a smart contract. Specifications cover a contract's properties and define how the contract should behave in different circumstances. The purpose of formal verification is to determine if a smart contract possesses these properties (invariants) and that these properties are not violated during execution. 

Formal specifications are critical in developing secure implementations of smart contracts. Contracts that fail to implement invariants or have their properties violated during execution are prone to vulnerabilities that can harm functionality or cause malicious exploits. 

## Types of formal specifications for smart contracts {#formal-specifications-for-smart-contracts}

Formal specifications enable mathematical reasoning about the correctness of program execution. As with formal models, formal specifications can capture either high-level properties or the low-level behavior of a contract implementation. 

> Formal specifications are derived using elements of [program logic](https://en.wikipedia.org/wiki/Logic_programming), which allow for formal reasoning about the properties of a program. A program logic has formal rules that express (in mathematical language) the expected behavior of a program. Various program logics are used in creating formal specifications, including [reachability logic](https://en.wikipedia.org/wiki/Reachability_problem), temporal logic, and Hoare logic. 

Reviewer 5: ```suggestion
Formal specifications are derived using elements of [program logic](https://en.wikipedia.org/wiki/Logic_programming), which allow for formal reasoning about the properties of a program. A program logic has formal rules that express (in mathematical language) the expected behavior of a program. Various program logics are used in creating formal specifications, including [reachability logic](https://en.wikipedia.org/wiki/Reachability_problem), [temporal logic](https://en.wikipedia.org/wiki/Temporal_logic), and [Hoare logic](https://en.wikipedia.org/wiki/Hoare_logic). 
```
I think we should include links on the rest of these like with reachability logic.

Reviewer 5: Ah I see you link out to this later. Still might be good to include here as well for consistency.

With formal verification, the question of verifying if a contract's business logic satisfies requirements is a mathematical proposition that can be proved or disproved. By formally proving a proposition, we can verify an infinite number of test cases with a finite number of steps. In this manner formal verification has better prospects of proving a contract is functionally correct with respect to a specification. 

#### Ideal verification targets {#ideal-verification-targets}

A verification target describes the system to be formally verified. Formal verification is best used in "embedded systems" (small, simple pieces of software that form part of a larger system). They are also ideal for specialized domains that have few rules, as this makes it easier to modify tools for verifying domain-specific properties. 

Smart contracts—at least, to some extent—fulfill both requirements. For example, the small size of Ethereum contracts makes them amenable to formal verification. Similarly, the EVM follows simple rules, which makes specifying and verifying semantic properties for programs running in the EVM easier. 

### Faster development cycle {#faster-development-cycle}

> Formal verification techniques, such as model checking and symbolic execution, are generally more efficient than regular analysis of smart contract code (performed during testing or auditing). This is because formal verification relies on symbolic values to test assertions("what if a user tries to withdraw *n* ether?") unlike testing which uses concrete values ("what if a user tries to withdraw 5 ether?"). 

Reviewer 5: ```suggestion
Formal verification techniques, such as model checking and symbolic execution, are generally more efficient than regular analysis of smart contract code (performed during testing or auditing). This is because formal verification relies on symbolic values to test assertions ("what if a user tries to withdraw *n* ether?") unlike testing which uses concrete values ("what if a user tries to withdraw 5 ether?"). 
```

- [GitHub](https://github.com/ethereum/act)
- [Documentation](https://ethereum.github.io/act/)

**Scribble** - _*Scribble transforms code annotations in the Scribble specification language into concrete assertions that check the specification.*_

- [Documentation](https://docs.scribble.codes/)

**Dafny** - _*Dafny is a verification-ready programming language that relies on high-level annotations to reason about and prove correctness of code.*_

- [GitHub](https://github.com/dafny-lang/dafny)

### Program verifiers for checking correctness {#program-verifiers}

**Solidity SMTChecker** - *_Solidity’s SMTChecker is a built-in model checker based on SMT (Satisfiability Modulo Theories) and Horn solving. It confirms if a contract’s source code matches specifications during compilation and statically checks for violations of safety properties._*

- [GitHub](https://github.com/ethereum/solidity)

**solc-verify** - *_solc-verify is an extended version of the Solidity compiler that can perform automated formal verification on Solidity code using annotations and modular program verification._*

- [GitHub](https://github.com/SRI-CSL/solidity)

> **hevm** - *_hevm is a symbolic execution engine and equivalence checker for EVM bytecode._*

Reviewer 3: I think hevm fits the symbolic execution section better.

- [GitHub](https://github.com/dapphub/dapptools/tree/master/src/hevm)

**KEVM**: _*KEVM is a formal semantics of the Ethereum Virtual Machine (EVM) written in the K framework. KEVM is executable and can prove certain property-related assertions using reachability logic.*_

- [GitHub](https://github.com/runtimeverification/evm-semantics)
- [Documentation](https://jellopaper.org/)

### Symbolic execution-based tools for detecting vulnerable patterns in smart contracts {#symbolic-execution-tools}

**Manticore** - *_A tool for analyzing EVM bytecode analysis tool based on symbolic execution_.*

- [GitHub](https://github.com/trailofbits/manticore)
- [Documentation](https://github.com/trailofbits/manticore/wiki)

**Securify** - _*A security analyzer for analyzing Solidity source code in Ethereum smart contracts and detecting property violations.*_

- [GitHub](https://github.com/eth-sri/securify2)

**Oyente** - *_Static analysis tool for analyzing vulnerabilities in EVM bytecode with symbolic execution._*

- [GitHub](https://github.com/melonproject/oyente)

> 

Reviewer 3: I'd remove Oyente and Securify.

Author: Removed them now.

**Echidna** - *_Echidna is a fast contract fuzzer for detecting vulnerabilities in smart contracts through property-based testing._*

- [GitHub](https://github.com/crytic/echidna/)

**Harvey** - *_Harvey is an automated fuzzing tool useful for detecting property violations in smart contract code._*

- [Website](https://consensys.net/diligence/fuzzing/)

> 

Reviewer 3: Echidna and Harvey are fuzzers, so maybe it'd make sense to have a fuzzers section even though this is an FV page? Or maybe move that to the testing page actually?

Author: That makes sense. I took them out now.

## Further reading {#further-reading}

### Articles {#articles}

- [How Formal Verification of Smart Contracts Works](https://runtimeverification.com/blog/how-formal-verification-of-smart-contracts-works/) [How Formal Verification Can Ensure Flawless Smart Contracts](https://media.consensys.net/how-formal-verification-can-ensure-flawless-smart-contracts-cbda8ad99bd1)
- [An Overview of Formal Verification Projects in the Ethereum Ecosystem](https://github.com/leonardoalt/ethereum_formal_verification_overview)

> - [An Introduction to Formal Verification of Ethereum Contracts](https://github.com/pirapira/Ethereum-Formal-Verification-Overview)

Reviewer 3: This repo hasn't been updated in 4 years, so might be a bit misleading.

- [SMTChecker and Formal Verification](https://docs.soliditylang.org/en/v0.8.15/smtchecker.html)

> 

Reviewer 3: Maybe worth adding https://runtimeverification.com/blog/end-to-end-formal-verification-of-ethereum-2-0-deposit-smart-contract/ as well?

Author: Added this too.
